docs/madr/decisions/011-rate-limit-policy.md

```diff
@@ -0,0 +1,358 @@
+resource for managing rate limiting.
+ 
+Rate Limiting in Kuma is implemented with Envoy's [rate limiting
+support](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter).
+There is an issue with the configuration of TCP rate limiting because [current filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/local_rate_limit_filter#config-network-filters-local-rate-limit) works on listener level which can only
+allow limiting connections on the specific listener but without checking the origin of the request.
+ 
+## Considered Options
+ 
+* Create a MeshRateLimit with basic TCP rate limit support
+* Create a MeshHTTPRateLimit and MeshTCPRateLimit with basic TCP rate limit support
+* Create advanced TCP rate limiting per requested service
+
+## Decision Outcome
+ 
+Chosen option: create a MeshRateLimit with basic TCP rate limit support
+ 
+## Decision Drivers
+ 
+- Replace existing policies with new policy-matching compliant resources
+- Add support for TCP connection limiting at the basic level
+- Possibility to add global rate limiting in the future
+ 
+## Considered Options
+ 
+### Naming
+ 
+- **MeshRateLimit**
+- MeshHTTPRateLimit and MeshTCPRateLimit
+ 
+## Solution
+ 
+### Current configuration
+Below is a sample rate-limiting configuration.
+ 
+```yaml
+spec:
+ sources:
+   - match:
+       kuma.io/service: "redis_kuma-demo_svc_6379"   
+ destinations:
+   - match:
+       kuma.io/service: "demo-app_kuma-demo_svc_5000"
+ conf:
+   http:
+     requests: 5
+     interval: 10s
+     onRateLimit:
+       status: 423
+       headers:
+         - key: "x-Kuma-rate-limited"
+           value: "true"
+           append: true
+```
+We are using [Envoy's local rate limit](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter) which supports only HTTP configuration.
+ 
+The configuration translates to the Envoy configuration on the route and the listener filter:
+ 
+```yaml
+{
+ "match": {
+   "prefix": "/",
+   "headers": [
+   {
+     "name": "x-kuma-tags",
+     "safe_regex_match": {
+     "google_re2": {},
+     "regex": ".*&kuma.io/service=[^&]*redis_kuma-demo_svc_6379[,&].*"
+     }
+   }
+   ]
+ },
+ "route": {
+   "cluster": "localhost:5000",
+   "timeout": "0s"
+ },
+ "typed_per_filter_config": {
+   "envoy.filters.http.local_ratelimit": {
+   "@type": "type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit",
+   ...
+   }
+ }
+}
+ 
+```
+ 
+### Specification
+ 
+The differences between TCP and HTTP rate limiting in `targetRef` and `from` requires different validation of objects.
+Idea is to create one policy but and allow configuring TCP and HTTP rate limiting at once. The only limitation is that TCP rate limit can be configured only when `from` section has `Mesh` and `targetRef` can have following values `Mesh|MeshSubset|MeshService|MeshServiceSubset`.
```

Comment:
```suggestion
We create one policy which can configure both TCP and HTTP rate limiting. Because, we can't identify incoming traffic when using TCP users can only set the TCP section when using the `Mesh` type of targetRef in the `from` section. For http these types of targetRef can be used: `Mesh|MeshSubset|MeshService|MeshServiceSubset`.
```